src/servicePrincipalManager.ts

```diff
@@ -200,3 +206,136 @@ export async function assignGroups({
     }
   }
 }
+
+export async function enableSaml({
+  displayName,
+  token,
+  objectId,
+  appId,
+}: {
+  displayName: string;
+  objectId: string;
+  token: string;
+  appId: string;
+}) {
+  const result = await fetch(
+    `https://graph.microsoft.com/v1.0/servicePrincipals/${objectId}`,
+    {
+      method: "PATCH",
+      headers: {
+        Authorization: `Bearer ${token}`,
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        preferredSingleSignOnMode: "saml",
+      }),
+    }
+  );
+
+  await errorHandler("Enabling Saml config", result);
+
+  await addTokenSigningCertificate({ displayName, token, objectId, appId });
+}
+
+async function addTokenSigningCertificate({
+  displayName,
+  token,
+  objectId,
+  appId,
+}: {
+  displayName: string;
+  objectId: string;
+  token: string;
+  appId: string;
+}) {
+  const servicePrincipal = await readServicePrincipal({
+    token,
+    servicePrincipalObjectId: objectId,
+  });
+  console.log(servicePrincipal);
+
+  //Adds a new signing certificate if all certificates are expiring.
+  if (
+    servicePrincipal.keyCredentials === undefined ||
+    servicePrincipal.keyCredentials.length == 0 ||
+    areAllCertficatesExpiring(servicePrincipal.keyCredentials)
+  ) {
+    const addCertificateResult = await createNewSigningCert(
+      objectId,
+      token,
+      displayName
+    );
+
+    await makeCertDefault(
+      objectId,
+      token,
+      (
+        await addCertificateResult.json()
+      ).thumbprint
```

Comment:
i think would be clearer if this was extracted to a variable, inline async looks a bit weird